### PR TITLE
[SPARK-12997][SQL] Use cast expression to perform type cast in csv

### PR DESCRIPTION
**Pull request by HyukjinKwon**
https://issues.apache.org/jira/browse/SPARK-12997
CSVTypeCast.castTo should probably be removed, and just replace its usage with a projection that uses a sequence of Cast expressions.

This PR resamples [SimpleTextRelation.scala#L115-L121](https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/sources/SimpleTextRelation.scala#L115-L121). 

For backword compatibility, it checks `,` for `DecimalType` and remove those for casting.

In addition, I renamed `csv.CSVInferSchema` to `csv.InferSchema` as JSON datasource has `json.InferSchema`. Although they have the same names, I did this because I thought the parent package name can still differentiate each.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@rxin @falaki Could you please look through this and tell me if I understood this correctly and this approach is appropriate?

**Comment by HyukjinKwon** [comment]
@rxin @falaki Could you please look through this and tell me if I understood this correctly and this approach is appropriate?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 1]
```diff
@@ -226,11 +227,17 @@ object CSVRelation extends Logging {
...
-            rowArray(subIndex) = CSVTypeCast.castTo(
-              indexSafeTokens(index),
-              field.dataType,
-              field.nullable,
-              params.nullValue)
+            val value = indexSafeTokens(index)
+            val dataType = field.dataType
+            val isNull =
+              value == params.nullValue && field.nullable && !dataType.isInstanceOf[StringType]
+            val safeValue = dataType match {
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Is this logic unit tested? I see you removed the unit test related to `DecimalType`. I suggest you move this to a utility function and unit test it. Would also be good to make sure we have one test in CSVSuite as well.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVTypeCastSuite.scala [hunk 2]
```diff
@@ -59,45 +49,4 @@ class CSVTypeCastSuite extends SparkFunSuite {
     }
     assert(exception.getMessage.contains("Unsupported special character for delimiter"))
   }
-
-  test("Nullable types are handled") {
```
Comment (falaki) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVTypeCastSuite.scala]: I think it is better to keep these tests. I suggest you add a function to `CSVRelation` companion object for type casting and then keep these unit tests.

**Comment by HyukjinKwon** [comment]
Hm.. I can't find a proper way to use `Cast()` for `DoubleType`, `FloatType` and `DecimalType`. Original ways of CSV casting works differently with the `Cast()` for those types.

I am a bit confused if I should keep this way or just just move back to the original codes and close this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 3]
```diff
@@ -33,7 +33,8 @@ import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat
 import org.apache.spark.Logging
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql._
-import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.catalyst.expressions.{Cast, Literal}
```
Comment (falaki) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Is this still needed?

**Comment by falaki** [comment]
With latest changes, what are we trying to achieve with this?

**Comment by HyukjinKwon** [comment]
@falaki As far as I understood, I am trying to use the existing casting tools in Spark by using`Cast()`.
However, it ended up with additional cases for the types, `DecimalType`,`FloatType` and `DoubleType` since `Cast()` handles differently those types.

I am not too sure if we should share this since this datasource handles the casting differently (also for JSON datasource, it converts data within `JacksonParser` itself without sharing this).

**Comment by HyukjinKwon** [comment]
@falaki Do you think we should close this?

**Comment by falaki** [comment]
If you think using `Cast()` does not work, yes we can close this.

**Comment by HyukjinKwon** [comment]
I am closing this.

BTW, I think the issue itself is a little bit questionable at the end. If we can apply the same rules for casting, then it would be great to re-use the `Cast()` but both CSV and JSON data sources are dealing with data types differently. 

As long as we are not going to this behaviour for backwards compatibility, I believe this issue is invalid or this will end up with some additional logics like I just did in this PR with some cases to deal with the differences.

**Pull request closed without merging**